Add support for extensions to the HTTP protocol

omarzouk opened the pull request:
<!--- Provide a general summary of your changes in the title above. -->

<!--- This template is entirely optional and can be removed, but is here to help both you and us. -->
<!--- Anything on lines wrapped in comments like these will not show up in the final text. -->

## Description
The GQL spec specifies a field called extensions as part of the [request parameters spec](https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#request-parameters) in the HTTP protocol. We are missing support for this

<!--- Describe your changes in detail here. -->
Adding the field to be parsed and propagating it as part of the `Info` object. This is done by wrapping the entire context object in a temporary container, and then unwrapping it inside the `Info` context getter. A new function is also added to the `Info` class called `input_extensions`. It exposes the extensions data.

Example usage by the end-user:
```python
@strawberry.field
def value_from_extensions(self, key: str, info: strawberry.Info) -> str:
    return info.input_extensions[key]
```

## Types of Changes

<!--- What types of changes does your pull request introduce? Put an `x` in all the boxes that apply. -->
- [X] Core
- [ ] Bugfix
- [X] New feature
- [ ] Enhancement/optimization
- [X] Documentation

## Issues Fixed or Closed by This PR

* fixes https://github.com/strawberry-graphql/strawberry/issues/3224

## Checklist

<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
- [x] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

<!-- Generated by sourcery-ai[bot]: start summary -->

## Summary by Sourcery

This pull request introduces support for the 'extensions' field in HTTP requests, allowing it to be parsed and included in the ExecutionContext. This change is accompanied by updates to the documentation and new tests to ensure proper functionality.

- **New Features**:
    - Added support for the 'extensions' field in HTTP requests, allowing it to be parsed and propagated as part of the ExecutionContext.
- **Documentation**:
    - Updated documentation to reflect the new support for the 'extensions' field in HTTP requests.
- **Tests**:
    - Added tests to verify the correct handling and propagation of the 'extensions' field in both GET and POST HTTP requests.

<!-- Generated by sourcery-ai[bot]: end summary -->

botberry commented:
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Omar Marzouk for the PR 👏

Get it here 👉 https://strawberry.rocks/release/(next)
```
<!-- action-check: release-file -->

botberry commented:
Hi, thanks for contributing to Strawberry 🍓!

We noticed that this PR is missing a `RELEASE.md` file. We use that to automatically do releases here on GitHub and, most importantly, to PyPI!

So as soon as this PR is merged, a release will be made 🚀.

Here's an example of `RELEASE.md`:

```markdown
Release type: patch

Description of the changes, ideally with some examples, if adding a new feature.
```

Release type can be one of patch, minor or major. We use [semver](https://semver.org/), so make sure to pick the appropriate type. If in doubt feel free to ask :)


Here's the tweet text: 
```
🆕 Release (next) is out! Thanks to Omar Marzouk for the PR 👏

Get it here 👉 https://strawberry.rocks/release/(next)
```
<!-- action-check: release-file -->

erikwrede reviewed (commented):
Thanks for the PR! Can you add a test to ensure the exension dict is actually correctly passed down the chain? 🙂

omarzouk commented:
@erikwrede Hi! yea I think I need to do a fresh take on this, I think I hit a dead end with my current approach, I've been investigating how to do it, and I think we have a couple options:

1- Wrap the context field in another object that holds the extensions and the context, i.e. wrap the field before [this line](https://github.com/strawberry-graphql/strawberry/pull/3461), then unwrap it inside the [Info class](https://github.com/strawberry-graphql/strawberry/blob/main/strawberry/types/info.py#L82). We would then introduce an `input_extensions` or `protocol_extensions` field to Info that reads the extensions from the wrapper.

2- Extend the definition of the [get_context](https://github.com/strawberry-graphql/strawberry/blob/main/strawberry/http/async_base_view.py#L72) function, add a 3rd parameter called `extensions`. Usually http users use the `Request` param to get headers, extensions can be extracted by the user here as well. However, this would require moving the body parsing and data fetching to happen before the `get_context` is called, i.e. [this operation](https://github.com/strawberry-graphql/strawberry/blob/39ccffc65f6fc3dc7af69596e85d2aaf69b070af/strawberry/http/async_base_view.py#L91) would have to move above [this line](https://github.com/strawberry-graphql/strawberry/blob/main/strawberry/http/async_base_view.py#L161)

The benefit of doing Option 1 would be that we can add this right away to http, without having to modify other protocols, and just include a type check for the wrapper in the `info.context` getter. It can then be added to other protocols over time, in other PRs.

What do u think? should I go ahead and do Option 1?

omarzouk commented:
hi @erikwrede, I went ahead and implemented Option 1 mentioned above, which is to wrap the context object and unwrap it in the Info getter. I also added a new test that is passing successfully. Please let me know what you think. 🙏

omarzouk commented:
pinging @patrick91 as well, since we discussed this originally in https://github.com/strawberry-graphql/strawberry/issues/3224

bellini666 commented on strawberry/http/__init__.py:
```diff
@@ -33,6 +33,7 @@ class GraphQLRequestData:
     query: Optional[str]
     variables: Optional[Dict[str, Any]]
     operation_name: Optional[str]
+    extensions: Optional[Dict[str, Any]]
```
Wondering if this should default to `None`?

```suggestion
    extensions: Optional[Dict[str, Any]] = None
```

omarzouk replied:
@bellini666 Since the `variables` parameter in the same object is also optional, and didn't have a default value, I wanted to keep this consistent so I don't introduce unexpected behaviour in case it gets initialised automatically some other way? 🤷‍♂️

bellini666 replied:
Fair point!

Maybe all of them should? =P
I'm fine in keeping the consistency or changing everything, lets see what others think

bellini666 commented on strawberry/http/async_base_view.py:
```diff
@@ -22,6 +22,7 @@
 from strawberry.types import ExecutionResult
 from strawberry.types.graphql import OperationType
 
+from ..types.context_wrapper import ContextWrapper
```
I think it is ok (and usually nice) to use relative imports for something already inside it (i.e. using `from .<something>`), but when importing from a parent (i.e. `from ..<something>) I think it starts to get messy.

In this case I would suggest importing from `strawberry.types.context_wrapper`

omarzouk replied:
done

bellini666 commented on strawberry/http/sync_base_view.py:
```diff
@@ -23,6 +23,7 @@
 from strawberry.types import ExecutionResult
 from strawberry.types.graphql import OperationType
 
+from ..types.context_wrapper import ContextWrapper
```
Ditto

omarzouk replied:
done

bellini666 commented on strawberry/types/info.py:
```diff
@@ -79,8 +80,16 @@ def selected_fields(self) -> List[Selection]:
 
     @property
     def context(self) -> ContextType:
+        if type(self._raw_info.context) is ContextWrapper:
```
better?

```suggestion
        if isinstance(self._raw_info.context, ContextWrapper):
```

bellini666 commented on strawberry/types/info.py:
```diff
@@ -79,8 +80,16 @@ def selected_fields(self) -> List[Selection]:
 
+    @property
+    def input_extensions(self) -> Dict[str, Any]:
+        if type(self._raw_info.context) is ContextWrapper:
```
```suggestion
        if isinstance(self._raw_info.context, ContextWrapper):
```

bellini666 commented on strawberry/types/context_wrapper.py:
```diff
@@ -0,0 +1,8 @@
+@dataclass
+class ContextWrapper:
+    context: Optional[Any]
+    extensions: Optional[Dict[str, Any]]
```
Same suggestion as above

```suggestion
    extensions: Optional[Dict[str, Any]] = None
```

bellini666 commented on tests/views/schema.py:
```diff
@@ -108,6 +108,10 @@ def root_name(self) -> str:
         return info.context["custom_value"]
 
+    @strawberry.field
+    def value_from_extensions(self, key: str, info: strawberry.Info) -> str:
```
nit: info first, to follow the pattern of others that receive `info` as the first one, but this one receives an extra `key`

Maybe also adding a `, *, ` in the middle to enforce `key` to be passed as a keyword would be good?

```suggestion
    def value_from_extensions(self, info: strawberry.Info, key: str) -> str:
```

bellini666 commented on strawberry/http/base.py:
```diff
@@ -67,6 +67,15 @@ def parse_query_params(
+        if "extensions" in params:
+            extensions = params["extensions"]
+
+            if isinstance(extensions, list):
+                extensions = extensions[0]
```
Why only getting the first on the list?

From the rest of the code `extensions` should be a dict. When can it be a list? If it is, is there a possibility of it having more than one value inside it? If yes, what does that mean? Should we warn/error out about that?

omarzouk replied:
Since it has the exact same type definition as the `variables` parameter, I just copy pasted how that is being parsed. I think it might be a list in GET requests?

bellini666 replied:
Hrmm I see. @strawberry-graphql/core do you know why we do that?

DoctorJohn replied:
I just dug around a little bit. Before #2681 was merged there used to be the following comment in the Sanic integration:

https://github.com/strawberry-graphql/strawberry/blob/860e5f236b936440d4d28b8cea3f7bc9316a1612/strawberry/sanic/views.py#L101-L108

This could explain why we parse variables in that way.

DoctorJohn replied:
Update: I'm working on a PR for this. Looks like it might be Sanics fault after all.

DoctorJohn replied:
#3558

bellini666 reviewed (commented):
Some comments, but I'm not totally aware of this, so other @strawberry-graphql/core will be able to do a better review :)

omarzouk commented:
Hello! apart from the extra check for list in the parameter parsing which is potentially redundant and might be removed, as I understand it, that change is also not really blocking for this PR, right? https://github.com/strawberry-graphql/strawberry/pull/3558 could be updated to remove the list parsing for `extensions` after this one is merged?

Is there any other comments regarding this PR? the approach in general around wrapping context and exposing the extra field in the `Info` object? how would you like me to proceed?

@bellini666 @DoctorJohn @erikwrede @patrick91

omarzouk commented:
Hi! I have merged main to include https://github.com/strawberry-graphql/strawberry/pull/3558 and removed the extra list type check for the extensions param. Anything else you guys would like me to change about this PR? if not I can should I create a release file?

@bellini666 @DoctorJohn @erikwrede @patrick91

patrick91 commented on strawberry/types/info.py:
```diff
@@ -79,8 +80,16 @@ def selected_fields(self) -> List[Selection]:
+            return self._raw_info.context.context
         return self._raw_info.context
 
+    @property
+    def input_extensions(self) -> Dict[str, Any]:
+        if isinstance(self._raw_info.context, ContextWrapper):
+            return self._raw_info.context.extensions
+        return {}
```
how come this is called input_extensions? 😊

omarzouk replied:
honestly, I just didn't have a better name 😅  I just felt `extensions` is used in so many contexts in strawberry and people might be confused as to which one this is. But I'm happy to change it to whatever you think makes most sense

patrick91 commented on strawberry/http/sync_base_view.py:
```diff
@@ -112,11 +113,15 @@ def execute_operation(
 
         assert self.schema
 
+        context_wrapper = ContextWrapper(
+            context=context, extensions=request_data.extensions
+        )
```
is this context wrapper only to pass request data?

omarzouk replied:
yes exactly. Since the `context` data comes from `get_context` and that can be any type of object the user decides to return, this felt like a safe way to attach more data

patrick91 replied:
I'll try to review and merge this this week 😊 thanks for the patience!

omarzouk commented:
hello! sorry to keep pinging, we are really eager to use this in our APIs, do u think its possible to get some traction on it? I'm happy to help any way I can @patrick91 🙏

patrick91 commented:
@omarzouk sure, I think I'll have time this week(end) (I'm currently travelling) 😊

And please consider sponsoring us, so we can prioritise issues better ;)

patrick91 commented:
@omarzouk an update on this, I think I found a nice way to pass the request extensions to info, but it is quite different from this, so I might try it in another PR

omarzouk commented:
> And please consider sponsoring us, so we can prioritise issues better ;)

@patrick91 Hi Patrick! I talked with our management about  it, it took a while because github's invoice system for sponsorships is a big mess, but we [Soundtrack Technologies](https://github.com/soundtrackyourbrand) is now a 🍓 Sponsor! you guys built an awesome library, we use it for a lot of our apps, so it makes all the sense in the world to help you maintain it!

patrick91 commented:
@omarzouk looks there's an alpha version of GraphQl-core which makes passing data around much easier, would you be ok with using their alpha?

omarzouk commented:
hi @patrick91 what is the status of this now? I see graphql-core have implemented some support for it in the [alpha](https://github.com/graphql-python/graphql-core/issues/226#issuecomment-2629309490) version. Are we working on something in Strawberry so we can make use of the new capability?

patrick91 commented:
@omarzouk I'll take a look at this tomorrow! sorry!

patrick91 commented:
/pre-release

patrick91 commented:
@omarzouk I think we (@bellini666) got a nice implementation, would be interested in trying the pre-release? 😊

botberry commented:
# Pre-release
:wave:

Releasing commit [832ad16db4024690c2bf1e27220c0ca2f1a1374a] to PyPi as pre-release! :package:

botberry commented:
# Pre-release
:wave:

Pre-release **0.269.0.dev.1747164009** [832ad16db4024690c2bf1e27220c0ca2f1a1374a] has been released on PyPi! :rocket:
You can try it by doing:

```shell
poetry add strawberry-graphql==0.269.0.dev.1747164009
```

omarzouk commented on strawberry/types/info.py:
```diff
@@ -113,6 +113,11 @@ def context(self) -> ContextType:
         return self._raw_info.context
 
+    @property
+    def input_extensions(self) -> dict[str, Any]:
```
@patrick91 First of all, thanks a lot for taking the time to do it! I tested it in one of our APIs, the data does go through all the way! However it is returned as a string and not as a dict, so we need to `json.loads` the data before we return it from this function. The [spec](https://github.com/graphql/graphql-over-http/blob/main/spec/GraphQLOverHTTP.md#request-parameters) states that the parameter is expected to be a map so parsing it should be safe to do.

omarzouk replied:
@patrick91 actually nvm I was wrong, I was over-escaping the input, totally my bad. The implementation is actually correct!

patrick91 replied:
@omarzouk awesome! I'll release this after we fix the unrelated issue with GlobalID :D

omarzouk reviewed (commented):
The pre-release works perfectly, thanks a lot guys! ❤️

botberry commented:
Thanks for contributing to Strawberry! 🎉 You've been invited to join
the Strawberry GraphQL organisation 😊

You can also request a free sticker by filling this form: https://forms.gle/dmnfQUPoY5gZbVT67

And don't forget to join our discord server: https://strawberry.rocks/discord 🔥

mdgilene commented:
Input extension information only being available from the `Info` object seems pretty limiting. It would make sense that someone would want to implement a custom extension and have access to the input extension information from the `ExecutionContext`. But currently from what I can tell this is not the case currently.

Are there any plans to make this info available to the various extension lifecycle hook methods?

patrick91 commented:
@mdgilene can you open an issue with an example use case? We'll try to make it happen 😊
